test/controllers/client/ClientControllerIntegrationTest.java

```diff
@@ -0,0 +1,68 @@
+package controllers.client;
+
+import controllers.ClientController;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.jupiter.api.Assertions;
+import org.mockito.Mockito;
+import services.implementation.ClientServiceImpl;
+import services.interfaces.ClientService;
+import util.fixtures.ClientServiceFixtures;
+
+public class ClientControllerIntegrationTest {
+  private ClientController clientController;
+  private ClientService clientServiceSpy;
+
+  @Before
+  public void setUp() {
+    ClientService clientService = new ClientServiceImpl();
+    clientServiceSpy = Mockito.spy(clientService);
+    clientController = new ClientController(clientServiceSpy);
+  }
+
+  @Test
+  public void findById_ShouldReturnOptionalOfClient() {
+    var client = ClientServiceFixtures.buildClient();
+    clientController.add(client);
+    var clientFound = clientController.findById(client.getId());
+    Assertions.assertEquals(clientFound, client);
+    Mockito.verify(clientServiceSpy, Mockito.times(1)).findById(Mockito.eq(client.getId()));
+  }
+
+  @Test
+  public void findAll_ShouldReturnAllClients() {
+    var clients = ClientServiceFixtures.buildClients(3);
+    clients.stream().forEach(clientController::add);
+    var clientsFound = clientController.getAll();
+    Assertions.assertEquals(clients, clientsFound);
+  }
+
+  @Test
+  public void add_ShouldReturnClient() {
+    var client = ClientServiceFixtures.buildClient();
+    var clientInserted = clientController.add(client);
+    Assertions.assertEquals(clientInserted, client);
+    Mockito.verify(clientServiceSpy, Mockito.times(1)).insert(Mockito.eq(client));
+  }
+
+  @Test
+  public void delete_ShouldReturnDeletedClient() {
+    var clients = ClientServiceFixtures.buildClients(3);
+    clients.stream().forEach(clientController::add);
+    var deletedClient = clientController.deleteById(clients.get(0).getId());
+    Assertions.assertEquals(deletedClient, clients.get(0));
+    Mockito.verify(clientServiceSpy, Mockito.times(1)).delete(Mockito.eq(clients.get(0).getId()));
+  }
+
+  @Test
+  public void update_ShouldReturnUpdatedClient() {
+    var clients = ClientServiceFixtures.buildClients(3);
+    clients.stream().forEach(clientController::add);
+    var clientToUpdate = clients.get(0);
+    clientToUpdate.setName("DIANA MARIA");
+    clientToUpdate.setLastName("Monegro Diaz");
+    var updatedClient = clientController.update(clientToUpdate);
+    Assertions.assertEquals(updatedClient, clientToUpdate);
+    Mockito.verify(clientServiceSpy, Mockito.times(1)).update(Mockito.eq(clientToUpdate));
```

Comment:
Could you build a new client in order to test this properly?

Reply:
Done, did that in the update and delete methods of the test, since those are the ones that change properties of already existing objects